src/optional.rs

```diff
@@ -0,0 +1,253 @@
+/// assert_eq!(serialized, &[1, 9]);
+///
+/// let root = ssz.tree_hash_root();
+/// let equivalent_list: VariableList<u64, U1> = VariableList::from(vec![9; 1]);
+/// assert_eq!(root, equivalent_list.tree_hash_root());
+///
+/// // Create an `Optional` from an `Option` that is `None`.
+/// let none: Option<u8> = None;
+/// let ssz: Optional<u8> = Optional::from(none);
+/// let serialized: &[u8] = &ssz.as_ssz_bytes();
+/// let null: &[u8] = &[];
+/// assert_eq!(serialized, null);
+///
+/// let root = ssz.tree_hash_root();
+/// let equivalent_list: VariableList<u8, U0> = VariableList::from(vec![]);
+/// assert_eq!(root, equivalent_list.tree_hash_root());
+///
+/// ```
+#[derive(Debug, Clone, Serialize, Deserialize, Derivative)]
+#[derivative(PartialEq, Hash(bound = "T: std::hash::Hash"))]
+#[serde(transparent)]
+pub struct Optional<T> {
+    optional: Option<T>,
+}
+
+impl<T> From<Option<T>> for Optional<T> {
+    fn from(optional: Option<T>) -> Self {
+        Self { optional }
+    }
+}
+
+impl<T> From<Optional<T>> for Option<T> {
+    fn from(val: Optional<T>) -> Option<T> {
+        val.optional
+    }
+}
+
+impl<T> Default for Optional<T> {
+    fn default() -> Self {
+        Self { optional: None }
+    }
+}
+
+impl<T> tree_hash::TreeHash for Optional<T>
+where
+    T: tree_hash::TreeHash,
+{
+    fn tree_hash_type() -> tree_hash::TreeHashType {
+        tree_hash::TreeHashType::List
+    }
+
+    fn tree_hash_packed_encoding(&self) -> tree_hash::PackedEncoding {
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_packing_factor() -> usize {
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_root(&self) -> Hash256 {
+        let root = optional_tree_hash_root::<T>(&self.optional);
+
+        let length = match &self.optional {
+            None => 0,
+            Some(_) => 1,
+        };
+
+        tree_hash::mix_in_length(&root, length)
+    }
+}
+
+impl<T> ssz::Encode for Optional<T>
+where
+    T: ssz::Encode,
+{
+    fn is_ssz_fixed_len() -> bool {
+        T::is_ssz_fixed_len()
+    }
+
+    fn ssz_fixed_len() -> usize {
+        <T>::ssz_fixed_len() + 1
+    }
+
+    fn ssz_bytes_len(&self) -> usize {
+        match &self.optional {
+            None => 0,
+            Some(val) => val.ssz_bytes_len() + 1,
+        }
+    }
+
+    fn ssz_append(&self, buf: &mut Vec<u8>) {
+        match &self.optional {
+            None => (),
+            Some(val) => {
+                let mut optional_identifier = vec![1];
+                buf.append(&mut optional_identifier);
+                val.ssz_append(buf)
+            }
+        }
+    }
+}
+
+impl<T> ssz::Decode for Optional<T>
+where
+    T: ssz::Decode,
+{
+    fn is_ssz_fixed_len() -> bool {
+        false
+    }
+
+    fn from_ssz_bytes(bytes: &[u8]) -> Result<Self, ssz::DecodeError> {
+        if bytes.is_empty() {
+            Ok(Optional { optional: None })
+        } else if bytes[0] == 0x01 {
```

Comment:
Minor nit, but this could be written without indexing as `if let Some((first, rest)) = bytes.split_first() { ... } else { .. }`.